# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/CanvasCaptureMediaStream.cpp
# Commit: c54cb3c10992
# Full Hash: c54cb3c109922bc15fb6d0ca67e5f2a05980c1a5
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2019-07-31 15:46:43
# Regressor Bug: 1493613
# File Overlap Count: 2
# Description:
#   Bug 1493613 - Move MediaStream control from DOMMediaStream to MediaStreamTrack. r=padenot
#   
#   This is inherently large, because modifying these bits of DOMMediaStream and
#   MediaStreamTrack affects all consumers and producers of all DOMMediaStreams and
#   MediaStreamTracks.
# ==============================================================================

diff -r 0b03dd9d20ac -r c54cb3c10992 dom/media/CanvasCaptureMediaStream.cpp
--- a/dom/media/CanvasCaptureMediaStream.cpp	Wed Jul 31 09:35:06 2019 +0000
+++ b/dom/media/CanvasCaptureMediaStream.cpp	Wed Jul 31 07:58:17 2019 +0000
@@ -44,7 +44,9 @@
 
 void OutputStreamDriver::EndTrack() {
   MOZ_ASSERT(NS_IsMainThread());
-  mSourceStream->EndTrack(mTrackId);
+  if (!mSourceStream->IsDestroyed()) {
+    mSourceStream->Destroy();
+  }
 }
 
 void OutputStreamDriver::SetImage(const RefPtr<layers::Image>& aImage,
@@ -148,7 +150,7 @@
 
 CanvasCaptureMediaStream::CanvasCaptureMediaStream(nsPIDOMWindowInner* aWindow,
                                                    HTMLCanvasElement* aCanvas)
-    : DOMMediaStream(aWindow), mCanvas(aCanvas), mOutputStreamDriver(nullptr) {}
+    : DOMMediaStream(aWindow), mCanvas(aCanvas) {}
 
 CanvasCaptureMediaStream::~CanvasCaptureMediaStream() {
   if (mOutputStreamDriver) {
@@ -168,36 +170,26 @@
 }
 
 nsresult CanvasCaptureMediaStream::Init(const dom::Optional<double>& aFPS,
-                                        const TrackID& aTrackId,
+                                        const TrackID aTrackId,
                                         nsIPrincipal* aPrincipal) {
+  MediaStreamGraph* graph = MediaStreamGraph::GetInstance(
+      MediaStreamGraph::SYSTEM_THREAD_DRIVER, mWindow,
+      MediaStreamGraph::REQUEST_DEFAULT_SAMPLE_RATE);
+  SourceMediaStream* source = graph->CreateSourceStream();
   PrincipalHandle principalHandle = MakePrincipalHandle(aPrincipal);
-
   if (!aFPS.WasPassed()) {
-    mOutputStreamDriver = new AutoDriver(GetInputStream()->AsSourceStream(),
-                                         aTrackId, principalHandle);
+    mOutputStreamDriver = new AutoDriver(source, aTrackId, principalHandle);
   } else if (aFPS.Value() < 0) {
     return NS_ERROR_ILLEGAL_VALUE;
   } else {
     // Cap frame rate to 60 FPS for sanity
     double fps = std::min(60.0, aFPS.Value());
-    mOutputStreamDriver = new TimerDriver(GetInputStream()->AsSourceStream(),
-                                          fps, aTrackId, principalHandle);
+    mOutputStreamDriver =
+        new TimerDriver(source, fps, aTrackId, principalHandle);
   }
   return NS_OK;
 }
 
-already_AddRefed<CanvasCaptureMediaStream>
-CanvasCaptureMediaStream::CreateSourceStream(nsPIDOMWindowInner* aWindow,
-                                             HTMLCanvasElement* aCanvas) {
-  RefPtr<CanvasCaptureMediaStream> stream =
-      new CanvasCaptureMediaStream(aWindow, aCanvas);
-  MediaStreamGraph* graph = MediaStreamGraph::GetInstance(
-      MediaStreamGraph::SYSTEM_THREAD_DRIVER, aWindow,
-      MediaStreamGraph::REQUEST_DEFAULT_SAMPLE_RATE);
-  stream->InitSourceStream(graph);
-  return stream.forget();
-}
-
 FrameCaptureListener* CanvasCaptureMediaStream::FrameCaptureListener() {
   return mOutputStreamDriver;
 }
@@ -212,5 +204,12 @@
   mOutputStreamDriver = nullptr;
 }
 
+SourceMediaStream* CanvasCaptureMediaStream::GetSourceStream() const {
+  if (!mOutputStreamDriver) {
+    return nullptr;
+  }
+  return mOutputStreamDriver->mSourceStream;
+}
+
 }  // namespace dom
 }  // namespace mozilla